uniffi_bindgen/src/library_mode.rs

```diff
@@ -138,39 +142,73 @@ fn find_components(
     metadata_groups
         .into_values()
         .map(|group| {
-            let package = find_package_by_crate_name(cargo_metadata, &group.namespace.crate_name)?;
-            let mut ci = ComponentInterface::new(&group.namespace.crate_name);
-            if let Some(metadata) = udl_items.remove(&group.namespace.crate_name) {
+            let crate_name = &group.namespace.crate_name;
+            let lib_info = crates
+                .get_crate(crate_name)
+                .ok_or(anyhow::anyhow!("no metadata for crate {crate_name}"))?;
+            let mut ci = ComponentInterface::new(crate_name);
+            if let Some(metadata) = udl_items.remove(crate_name) {
                 ci.add_metadata(metadata)?;
             };
             ci.add_metadata(group)?;
-            Ok((ci, package))
+            Ok((ci, lib_info.clone()))
         })
         .collect()
 }
 
-fn find_package_by_crate_name(
-    metadata: &cargo_metadata::Metadata,
-    crate_name: &str,
-) -> Result<Package> {
-    let matching: Vec<&Package> = metadata
-        .packages
-        .iter()
-        .filter(|p| {
-            p.targets
-                .iter()
-                .any(|t| t.name.replace('-', "_") == crate_name)
-        })
-        .collect();
-    match matching.len() {
-        1 => Ok(matching[0].clone()),
-        n => bail!("cargo metadata returned {n} packages for crate name {crate_name}"),
+#[derive(Debug, Clone, Default)]
+struct LibMap {
+    libs: HashMap<String, LibInfo>,
+}
+
+impl LibMap {
+    fn get_crate(&self, crate_name: &str) -> Option<&LibInfo> {
+        self.libs.get(crate_name)
     }
 }
 
+#[cfg(feature = "cargo_metadata")]
+impl From<cargo_metadata::Metadata> for LibMap {
+    fn from(metadata: cargo_metadata::Metadata) -> Self {
+        let libs: HashMap<String, LibInfo> = metadata
+            .packages
+            .iter()
+            .flat_map(|p| {
+                p.targets.iter().filter(|t| t.is_lib()).map(|t| {
+                    (
+                        t.name.replace('-', "_"),
+                        LibInfo {
+                            package_name: p.name.clone(),
+                            path: p.manifest_path.parent().map(|p| p.to_owned()),
+                        },
+                    )
+                })
+            })
+            .collect();
+        Self { libs }
+    }
+}
+
+// We should think this struct through a little.
+// * package_name: used only for tests for swift - should we
```

Comment:
See https://github.com/mozilla/uniffi-rs/commit/19b639d4b40af963ff30e9caf9b5768d092b0730 - getting rid of `package_name` is actually quite easy and I think it's an improvement - it does leave this struct with 1 element though!